### ISSUE TITLE
Formatting of email responses in NewDot isn't pretty

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#8088]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
[@greeny](https://expensify.slack.com/team/U044D2MA9) was sending messages back/forth to a business partner

## Expected Result:
When the partner replied via email that the formatting would like nice and there wouldn't be extra line breaks

## Actual Result:
Formatting is a bit janky, inc. extra line breaks

## Workaround:
unknown 

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 1.1.41-3
**Reproducible in staging?:** need repro
**Reproducible in production?:** Y 
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 
<img width="1046" alt="image - 2022-03-11T000521 922" src="https://user-images.githubusercontent.com/43995119/157805706-78df4e7b-f7ba-4409-9e9b-4961cfab8df2.png">

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @mallenexpensify 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1646931708047189

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @MariaHCD (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by zsgreenwald** [comment]
On it!

**Comment by thienlnam** [comment]
Yeah, not much we can do here unless we try to strip it ourselves which leads to endless things we'd have to account for.  Since mailgun email stripping has been unreliable we can look into using another library for this except we'd need to test multiple platforms to make sure it works and isn't stripping out text content

**Comment by thienlnam** [comment]
Yup those options are correct, I think option 2 would actually be worthwhile to do

> Could you elaborate on those unintended changes? 

It's been a while so I might not be remembering correctly, but IIRC stripping whitespace messed up the email formatting inside the email and made (yahoo?) emails look especially bad.  I think we had this enabled at one point but decided to revert it.

> Maybe we can trim any whitespaces before and after the stripped HTML?

Yeah this could help a bit, though it looks like in the linked issue the spacing is bad between text and before/after the text too and so it would still look bad

**Comment by MariaHCD** [comment]
Great, thanks, @thienlnam! So moving forward, sounds like the best option would be to investigate usage of another library and this is something that can be handled internally as it is a backend change.

**Comment by MariaHCD** [comment]
> Triggered auto assignment to @NicMendonca (`Internal`) because issue was reported by a contributor who needs to be compensated if this issue is fixed

This issue was not reported by a contributor.

**Comment by aldo-expensify** [comment]
I worked in this in the past in [this PR](https://github.com/Expensify/Web-Expensify/pull/31965). I added a function to trim the html, but it missed to consider cases where there are spaces using `&nsbp;`. I'll do a PR to fix that.

**Comment by zsgreenwald** [comment]
@aldo-expensify @MariaHCD what's the latest with this? I'm still noticing some heavily spaced formatting on incoming messages from email. Some screenshots below:

<img width="937" alt="image" src="https://user-images.githubusercontent.com/11618505/167955341-2a73da0f-8e08-4af3-ab72-e1b5c0b84f7f.png">


<img width="940" alt="image" src="https://user-images.githubusercontent.com/11618505/167955348-84f4930f-95d2-4506-b3b5-605533432627.png">

**Comment by aldo-expensify** [comment]
Hey @zsgreenwald ,  I made a PR, but it seems like it didn't work in this case. I'll have a look and see what it may be

**Comment by zsgreenwald** [comment]
Thanks! Let me know if there's anything i can do to help

**Comment by aldo-expensify** [comment]
@zsgreenwald do you get these emails also in your email inbox? or just new dot? Having the email to test with would help :)

**Comment by zsgreenwald** [comment]
Hey @aldo-expensify what else do you need from me to push this forward? Thanks!

**Comment by aldo-expensify** [comment]
I'll put some time tomorrow to see if I can figure it out.

**Comment by quinthar** [comment]
Sorry for missing something -- what is the exact problem?  Is it extra whitespace?  Can someone describe specifically what they would like changed?

**Comment by zsgreenwald** [comment]
Yes, the exact problem is the extra white space. I'm suggesting we tighten it so it looks more natural when we receive email responses. 

![image](https://user-images.githubusercontent.com/11618505/168734054-0cf81ddb-55e1-42ee-b1f1-d445399d1761.png)

Not pressing, but I figured it'd be a quick fix.

**Comment by thienlnam** [comment]
Ah this has been a PITA in the past, we've had a lot of trouble parsing / cleaning specially formatted emails from outlook which I believe this is - for these types of emails we delegate the html stripping to mailgun.  In the past, we've just stripped all line breaks to solve this problem but it removed any line breaks intended within the email which also didn't look great

@aldo-expensify You could try removing all break tags before text content starts, but there's a lot of email formatting overhead that you would have to figure out.

**Comment by aldo-expensify** [comment]
It is indeed a PITA, but I think there is room from improvement.

I'm testing locally with a message with the following html (this is what we end up with after our API processes the email):

` <div class="WordSection1"> <p class="MsoNormal">Hi Zach, </p><p></p> <p class="MsoNormal"></p><p> </p> <p class="MsoNormal">Will tomorrow at 3 est. work?</p><p></p> <p class="MsoNormal"></p><p> </p> <p class="MsoNormal">Thanks,  Steve</p><p></p> <p class="MsoNormal"></p><p> </p> </div>Notice required by law: This e-mail may constitute an advertisement or solicitation under U.S. law, if its primary purpose is to advertise or promote a commercial product or service. You may choose not to receive advertising and promotional messages from Crowe  LLP (except for the crowe.com website, which tracks e-mail preferences through a separate process) at this e-mail address by forwarding this message to CroweUnsubscribe@crowe.com. If you do so, the sender of this message will be notified promptly. Our principal  postal address is 225 W Wacker Drive, Suite 2600, Chicago, IL 60602-4903.  This email message is from Crowe LLP or one of its subsidiaries and may contain privileged or confidential information  or other information exempt from disclosure under applicable law. If you are not the intended recipient, please notify the sender by reply email immediately and delete this message without reading further or forwarding to others. This email is not intended  to be a contract or other legally binding obligation, and any tax advice expressed in this email should not be construed as a formal tax opinion unless expressly stated. Visit www.crowe.com/disclosure for more information about Crowe LLP and its subsidiaries.     `

<img width="1381" alt="image" src="https://user-images.githubusercontent.com/87341702/168875130-ecc981a0-94cd-4cb9-a050-88742788211f.png">



The problem seems to be that whitespaces before the text "Hi Zach" are getting wrapped in divs. These are the white lines we see.

<img width="453" alt="image" src="https://user-images.githubusercontent.com/87341702/168875842-e1c22e34-8e83-4fa2-93b8-808e92ee4def.png">

become these divs:

<img width="1751" alt="image" src="https://user-images.githubusercontent.com/87341702/168876014-6aab1c5b-7e97-4cf9-94b4-7edcff0d8207.png">


I think this is a problem to fix in the front end, but I'm not sure where the html gets transformed yet.

On the backend, we have some code that tries to trim the html by removing tags that don't have text content and that don't have text content after or before them. Applying the same logic, I can remove the TEXT nodes that are just blank spaces in the borders of the email that are getting converted to divs (see this [draft PR](https://github.com/Expensify/Web-Expensify/pull/33790)). This works, but only for the white lines that are added at the top and bottom... but I think there are still some white lines added in between because of the spaces being wrapped in divs.


Result:

<img width="1377" alt="image" src="https://user-images.githubusercontent.com/87341702/168877029-6cdfd569-8823-4ec8-9db6-f64c7d6846bd.png">

**Comment by aldo-expensify** [comment]
Changing this prop [dangerouslyDisableWhitespaceCollapsing](https://github.com/Expensify/App/blob/b0b1cbea67704611b6032560a5b95e4be3afadf8/src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js#L56) to `false` seems to fix the problem of whitespaces getting wrapped in divs:


<img width="1398" alt="image" src="https://user-images.githubusercontent.com/87341702/168888027-cc087ad7-65f5-42ff-92e7-d84a3cb86654.png">

**Comment by aldo-expensify** [comment]
@parasharrajat this PR: https://github.com/Expensify/App/pull/6975 changed [dangerouslyDisableWhitespaceCollapsing](https://github.com/Expensify/App/blob/b0b1cbea67704611b6032560a5b95e4be3afadf8/src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js#L56) to `true`, which seems to be causing the problem here: unnecessary extra empty lines.

**Comment by parasharrajat** [comment]
We are removing that flag in another issue. I will add the PR link as soon as I find it.

**Comment by parasharrajat** [comment]
That PR is #9025

**Comment by aldo-expensify** [comment]
The solution in that PR seems to cause the exact same problem for this case:


https://user-images.githubusercontent.com/87341702/168906255-19b9f684-09a2-4a57-87ee-c60c3f83a171.mov



I don't think we should be changing how the whitespace works when rendering html outside a `<pre>`. If there is some indentation problems with lists.. maybe there is something to do with padding or margin?

**Comment by zsgreenwald** [comment]
@aldo-expensify I noticed [this PR](https://github.com/Expensify/App/pull/9025) merged. I'm waiting on some email responses to come in, but should that merge fix the issue?

**Comment by aldo-expensify** [comment]
> @aldo-expensify I noticed https://github.com/Expensify/App/pull/9025 merged. I'm waiting on some email responses to come in, but should that merge fix the issue?

I tested locally [that PR](https://github.com/Expensify/App/pull/9025) and it seems to cause the same problem.

**Comment by aldo-expensify** [comment]
I investigate this a bit more today and I know better now when it happens, but I haven't been able to come up with a solution. I have a clearer way to reproduce it here: https://github.com/Expensify/App/pull/9195#issuecomment-1146486983

**Comment by aldo-expensify** [comment]
From what I can see, we have the following cases:

<table>
<tr>
<th>
</th>
<th>
With <code>white-space: pre<code>
</th>
<th>
Without `white-space: pre`
</th>
<th width="200px">
Data
</th>
</tr>
<tr>
<td>
1
</td>
<td>
<img src="https://user-images.githubusercontent.com/87341702/172508166-fa1e2b7c-7646-4c72-b682-dd1d6df7bbdf.png">
</td>
<td>
<img src="https://user-images.githubusercontent.com/87341702/172508166-fa1e2b7c-7646-4c72-b682-dd1d6df7bbdf.png">
</td>
<td>
<code>
"0 spaces&lt;br /&gt; 1 spaces&lt;br /&gt;  2 spaces&lt;br /&gt;   3 spaces&lt;br /&gt;    4 spaces"
</code>
</td>
</tr>
<tr>
<td>
2
</td>
<td>
<img src="https://user-images.githubusercontent.com/87341702/172508753-ba503f86-18df-4d49-add5-b4dadf5744a5.png"
</td>
<td>
<img src="https://user-images.githubusercontent.com/87341702/172509687-6d404d8a-17b8-43cf-96b2-3aa6f44a6c07.png">
</td>
<td>
<code>
"0 spaces&lt;br /&gt; 1 spaces&lt;br /&gt;  2 spaces&lt;br /&gt;   3 spaces&lt;br /&gt;    4 spaces&lt;br /&gt; &lt;pre&gt;Test&lt;/pre&gt;"
</code>
</td>
</tr>
<tr>
<td>
3
</td>
<td>
<img src="https://user-images.githubusercontent.com/87341702/172508832-6b3591db-27aa-45c1-9905-e2a87cc9e189.png">
</td>
<td>
<img src="https://user-images.githubusercontent.com/87341702/172509741-baad2754-c185-4f62-8c16-5678cdc8a651.png">
</td>
<td>
"     &lt;div class=\&quot;WordSection1\&quot;&gt; &lt;p class=\&quot;MsoNormal\&quot;&gt;Hi SomeName, &lt;/p&gt;&lt;p&gt;&lt;/p&gt; &lt;p class=\&quot;MsoNormal\&quot;&gt;&lt;/p&gt;&lt;p&gt; &lt;/p&gt; &lt;p class=\&quot;MsoNormal\&quot;&gt;Will tomorrow at 34 est. work?&lt;/p&gt;&lt;p&gt;&lt;/p&gt; &lt;p class=\&quot;MsoNormal\&quot;&gt;&lt;/p&gt;&lt;p&gt; &lt;/p&gt; &lt;p class=\&quot;MsoNormal\&quot;&gt;Thanks,  Test&lt;/p&gt;&lt;p&gt;&lt;/p&gt; &lt;p class=\&quot;MsoNormal\&quot;&gt;&lt;/p&gt;&lt;p&gt; &lt;/p&gt; &lt;/div&gt;xxxxxx xxxxxxxx xx xxx: xxxx x-xxxx xxx xxxxxxxxxx xx xxxxxxxxxxxxx xx xxxxxxxxxxxx xxxxx x.x. xxx, xx xxx xxxxxxx xxxxxxx xx xx xxxxxxxxx xx xxxxxxx x xxx..."
</td>
</tr>
</tr>

</table>

1. This case is *NOT* rendered as HTML because of <a href="https://github.com/Expensify/App/blob/784d5384ae0abe772fea8c129733f081c384d423/src/pages/home/report/ReportActionItemFragment.js#L93-L98">this exception</a> in handling it, so it works in both cases.
2. and 3. There are other html tags than &lt;br /&gt;, so we render these cases using the [react-native-render-html](https://meliorence.github.io/react-native-render-html/docs/intro) library.

I think it makes sense to add the `white-space: pre` when we are rendering html messages typed by the user in our App because that makes them render as close as possible to what the user initially typed (preserving spacing). The problem appears when the message came from an email. The email client probably have a different way of rendering the HTML of the message, they probably don't use `white-space: pre`.

I was thinking that the safest bet would be to do something like:
- Flag the messages that came from the email intake
- In the front, don't force `white-space: pre` for the flagged messages

The problem I see with this approach is.. what if the user tries to edit the message that came from the email?

Another approach would be to try to do heavier processing of the email in the backend to avoid the amount of weird structured html in them, but that sounds like an endless amount of work.

**Comment by aldo-expensify** [comment]
Just mentioning that if you edit the message that came from the email reply, it "breaks" like this:

![image](https://user-images.githubusercontent.com/87341702/172706667-c3a9585c-168d-4ade-910f-eb74880670a9.png)

The result is the same with and without `white-space:pre`

**Comment by aldo-expensify** [comment]
I did some more testing today and found out that outlook likes to add a bunch of `\n` in the email's html. The `\n` with `whitespace: pre` makes a lot of empty lines appear.

The proposed solution of avoid using `whitespace: pre` for message comings from emails handles well this case, so I'm sticking with that.

I updated the PR's descriptions and made them ready for review.
